File: Data Structures & Algorithms/top-k-elements-in-list/submission-12.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        int count = 0;
        int[] res = new int[k];
        List<Integer>[] bucket = new ArrayList[nums.length+1];
        for (int i = 0; i < bucket.length; i++) {
            bucket[i] = new ArrayList<>();
        }
        Map<Integer, Integer> freq = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int num = nums[i];
            freq.put(nums[i], freq.getOrDefault(nums[i], 0) + 1);
        }
        for (int key : freq.keySet()) {
            bucket[freq.get(key)].add(key);
        }
        for (int i = bucket.length-1; i >= 0; i--) {
            if (!bucket[i].isEmpty()) {
                for (int e : bucket[i]) {
                    res[count] = e;
                    count++;
                }
                if (count == k) {
                    break;
                }
            }
        }
        return res;
    }
}

// 4 5 5 6 6 6
// num -> count
// 4 -> 1
// 5 -> 2
// 6 -> 3
// k = 2
// [5, 6]

// frequency map of all elements
// use a heap of size k
// o(n log k)

// vs bucket sort O(n)
// have a list of all elements at the index frequency

File: Data Structures & Algorithms/three-integer-sum/submission-12.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        i = 0
        while i < (len(nums)-2):
            j = i+1
            k = len(nums) - 1
            while j < k:
                total = nums[i]+nums[j]+nums[k]
                if total == 0:
                    res.append([nums[i],nums[j],nums[k]])
                if total <= 0:
                    j_num = nums[j]
                    while j<k and nums[j] == j_num:
                        j+=1
                if total >= 0:
                    k_num = nums[k]
                    while k>j and nums[k] == k_num:
                        k-=1
            i_num = nums[i]
            while i < (len(nums)-2) and nums[i] == i_num:
                i+=1
        return res
                    